### PR TITLE
Implement Stable annotations for arrays

### PR DESCRIPTION
**Pull request by gita-omr**
- if array is annotated as stable, treat its elements as stable
- implement similar to arrays with constant elements except that
  array element cannot be folded if its value is 0
- handle only one dimension for now, until we have a test case
  and a need for multi-dimensional case

### CODE REVIEW

**Comment by gita-omr** [comment]
Jenkins build all

**Review by jdmpapin — COMMENTED** [review]
It looks like this PR and eclipse/omr#6242 will currently require a coordinated merge, at least due to `SymbolReferenceTable::findOrCreateSymRefWithKnownObject()`, and possibly also due to `KnownObjectTable::getOrCreateIndex[At]()`, though it's hard to tell about the latter. Furthermore, the changes to the signatures of these methods seem to be entirely Java-specific (since `@Stable` is Java-specific)

I think that there is a straightforward modification that could address both of these points:
1. Leave `stableArrayRank` out of the signatures of those methods
2. Define a new method, e.g. `void addStableArray(Index koi, int32_t stableArrayRank)` in OpenJ9
3. Update callers that want to specify a nonzero `stableArrayRank` to additionally call the new method

At that point, there should no longer be any incompatibility introduced, and the only OMR change would be to add `isArrayWithStableElements()` and use it in VP. And that could perhaps be renamed to (slightly) more generically reflect its purpose in OMR, e.g. `isMutableArrayWithPossiblyConstantElements`

What do you think?

File: runtime/compiler/env/J9KnownObjectTable.cpp
Comment (gita-omr) [reply runtime/compiler/env/J9KnownObjectTable.cpp:128]: I was debating whether to make this support completely OpenJ9 specific or leave an opportunity for OMR to use it as well, similar to the arrays with constant elements.  I think I ensured that OMR can be merged first [1]
But I am fine with making it openj9 specific.

[1] https://github.com/eclipse/omr/pull/6242

File: runtime/compiler/env/J9KnownObjectTable.cpp
Comment (gita-omr) [reply runtime/compiler/env/J9KnownObjectTable.cpp:133]: Oh, good idea! I wanted to avoid the array growing when checking non-existing value. But checking `i < _stableArrayRanks.size() ` should do the work. Certainly, better to remove the vector then. Thanks!

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:1448]: @jdmpapin what do you think of using transformIndirectLoadChainImpl() for Stable arrays? I feel that we need to duplicate a lot of the same checks that it already does. The indirection will stop automatically at the first level, as we would like to happen for stable array elements.

**Comment by gita-omr** [comment]
@jdmpapin I think implemented all your suggestions(which are great). I also merged stable array handling with regular field folding since I felt that handling arrays with immutable elements is very similar except: (1) we don't have a special shadow for stable array elements, and (2) we need to check for value not being 0. Otherwise the logic is very similar.

**Comment by gita-omr** [comment]
Removed WIP since it does not depend on https://github.com/eclipse/omr/pull/6242 anymore.

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:551]: It looks like this condition and the one in `verifyFieldAccess()` both assume that `isBaseStableArray` implies that `symRef` is an array shadow, but nothing has checked that it is. I think it would be good to check near the beginning of `transformIndirectLoadChainImpl()`, e.g.
```
if (isBaseStableArray && !symRef->getSymbol()->isArrayShadowSymbol())
   return false;
```

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:578]: If there is a mismatch between the type of the load and the array element type (which can only happen when both are primitive types since `verifyFieldAccess()` has succeeded), or if the address is not naturally aligned, then we have to take special care because it's the array elements that are stable, not just any sequence of bytes that we care to load from within the array

For example, if we were to load a 32-bit integer from a stable array of bytes, and if that 32-bit integer were nonzero, it might still overlap byte elements with value zero that can be overwritten later

As another example, if we were to load a 16-bit integer from a stable array of 16-bit integers, but at an odd address, and if the loaded value were nonzero, it too might still overlap an element with value zero that can be overwritten later

The exact safety test here would be to look at each array element (based on the array element type) that overlaps the value to be loaded, and to give up if any such element is zero

However, I don't think we need that level of complexity (and certainly not at the moment). It should suffice to check that the type of the load (i.e. `symRef->getSymbol()->getDataType()`) is the same as the array element type, and that the address is naturally aligned

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:1527]: If this were done within the conditional just below, it would respect the `performTransformation()`, and `addStableArray()` could take the known object index as a parameter instead of redundantly calling `getOrCreateIndexAt()`

File: runtime/compiler/env/J9KnownObjectTable.cpp
Comment (jdmpapin) [inline runtime/compiler/env/J9KnownObjectTable.cpp:524]: The name `result` is misleading here

**Review by jdmpapin — CHANGES_REQUESTED** [review]

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:367]: typo: dimention → dimension

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:578]: Yeah, that's a good point. This is another thing that makes stable array handling different from that of array with constant elements! (sorry, maybe I missed that in your earlier comments and also forgot that this type of access can happen). Then, we will need to pass the type of the array element along with the `isBaseStableArray.` Although I still think it would be good to treat them as similar as possible.

Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:578]: Yes, you are right. This is another thing that make stable arrays different from arrays with constant elements. Let me see how I can add these checks.

File: runtime/compiler/env/J9KnownObjectTable.cpp
Comment (gita-omr) [reply runtime/compiler/env/J9KnownObjectTable.cpp]: I can't find any check whether the object is an array for addArrayWithConstantElements. Do you mean we should check for both?

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:578]: @jdmpapin, when you have a chance, could you please point me to a method that returns TR::DataType  for array element given array address?

Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:578]: Maybe we can use `J9::ObjectModel::getArrayElementWidthInBytes(TR::Compilation* comp, uintptr_t objectPointer)` and check if sizes of array elements and the symbol reference we are using to access them are the same? In addition we will check (fieldAddress%size ==0 ) ?

Comment (jdmpapin) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:578]: I think that last suggestion works. We already know from `verifyFieldAccess` that we're either loading a reference from a reference array or a primitive from a primitive array

In the reference case, the sizes will appear to mismatch with compressed references, but it will actually be ok

In the primitive case, the only pairs of distinct types with the same sizes are boolean/byte, short/char, int/float, and long/double. For boolean/byte and short/char, there is no ambiguity about whether a value is zero. For int/float and long/double, we can load an integer element as a floating point number or vice versa. If we load an integer element as floating point, it might appear to be -0.0 and fail to fold even though the element is now constant, but failure to fold is safe. If OTOH we load a floating point element as an integer, we will fold anything that has at least one bit set. The only floating point value that `isNullValueAtAddress()` would have us refuse to fold despite having a bit set is -0.0, but I think it should be safe to fold that value anyway, since it has definitely been modified after the array was zero-initialized

So *as long as it's ok to fold -0.0*, we just require that `fieldAddress` is aligned and the array either is a reference array or has an element size matching the array shadow symbol

Flipping that around, we need to prevent folding if `fieldAddress` is misaligned or the array is a primitive array whose element size does not match the array shadow symbol

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:196]: The types here were exhaustive for `avoidFoldingInstanceField()`, but for array element loads this function will also need to handle `Int8` and `Int16`

File: runtime/compiler/env/J9KnownObjectTable.cpp
Comment (jdmpapin) [reply runtime/compiler/env/J9KnownObjectTable.cpp]: I was only commenting about stable arrays, since those are what's new, but I do think it would make sense for `addArrayWithConstantElements()` as well

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:196]: You are absolutely right! But I am wondering why don't we check those types  in `J9::TransformUtil::transformIndirectLoadChainImpl() ` when we do the actual folding?

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:578]: Right, we might fold -0.0 if we read int as float... But I think it's consistent with JEP's idea of the field/array element becoming constant after second value... Also, Stable is only used inside JCL and unlikely is used in such an ambiguous way.

**Comment by gita-omr** [comment]
Applied the changes and ran the tests.

File: runtime/compiler/env/J9KnownObjectTable.cpp
Comment (gita-omr) [reply runtime/compiler/env/J9KnownObjectTable.cpp]: I added assert to only addStableArray to minimize the changes in this PR.

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp]: Added the check for size and alignment.

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp]: Improved.

File: runtime/compiler/env/J9KnownObjectTable.cpp
Comment (gita-omr) [reply runtime/compiler/env/J9KnownObjectTable.cpp]: Improved.

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp]: Fixed

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:210]: Added two cases in `avoidFoldingInstanceField()`

**Comment by gita-omr** [comment]
I think I missed one comment. Looking at it.

**Comment by gita-omr** [comment]
Addressed another comment and white spaces.

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:572]: Added the check.

File: runtime/compiler/optimizer/J9TransformUtil.cpp
Comment (jdmpapin) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:210]: Interesting. I guess folding 8- and 16-bit array elements hasn't been particularly important for performance. I see no reason we shouldn't fold them, though that's not something you necessarily need to change in this PR. OTOH it would be really easy to add them. So I'll leave that up to you

But even with the final transformation refusing to fold `Int8` and `Int16`, the current logic would allow for an 8- or 16-bit load from a stable array to reach `isNullValueAtAddress()`, since it won't have been ignored yet. So these cases are necessary here either way. Thanks for adding them!

Comment (gita-omr) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:210]: Thanks! Good to know! I would like to limit the changes in this PR.

**Review by jdmpapin — APPROVED** [review]

**Comment by jdmpapin** [comment]
There is another dimension-related typo in the commit message and initial PR comment. dimentional → dimensional

**Comment by gita-omr** [comment]
Fixed the typo.

**Comment by jdmpapin** [comment]
Jenkins test sanity all jdk17 depends eclipse/omr#6242

**Comment by jdmpapin** [comment]
Because eclipse/omr#6242 is merged, I don't think there's any point testing these OpenJ9 changes in isolation, so I'm testing the combination of the two PRs, and I'll hold off merging this one until 6242 promotes into openj9-omr:openj9

**Comment by pshipton** [comment]
You would need to use `depends eclipse/omr#master` after a PR is merged. Anyway I just promoted OMR.

Jenkins test sanity all jdk17

**Comment by gita-omr** [comment]
@jdmpapin , looks like "Pull request failures"  are due to trying to checkout 6242.

**Comment by pshipton** [comment]
It's the second PR build you want to look at, which all passed https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/1266/

**Pull request merged**
